felix/wireguard/bootstrap.go

```diff
@@ -27,58 +26,312 @@ import (
+
+	// Make a few attempts to read our publickey from the datastore, compare. If different, remove all wireguard
+	// configuration.
+	for r := 0; r < bootstrapMaxRetries; r++ {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		thisNode, err := calicoClient.Nodes().Get(ctx, nodeName, options.GetOptions{})
+		cancel()
+		if err != nil {
+			logCtx.WithError(err).Warn("Couldn't fetch node config from datastore, retrying")
+			<-expBackoffMgr.Backoff().C() // safe to block here as we're not dependent on other threads
+			continue
+		}
+
+		// If there is any config mismatch, wipe the datastore's publickey (forces peers to send unencrypted traffic).
+		// Once connected and sync'd the key will be regenerated and propagated.
+		storedPublicKey = thisNode.Status.WireguardPublicKey
+		if storedPublicKey != kernelPublicKey {
+			logCtx.Info("Found mismatch between kernel and datastore wireguard keys - removing wireguard configuration")
+			return RemoveWireguardForHostEncryptionBootstrapping(configParams, getNetlinkHandle, calicoClient)
+		}
+		return nil
+	}
+
+	// Couldn't determine the current wireguard configuration.
+	return fmt.Errorf("couldn't determine current wireguard configuration after %d retries", bootstrapMaxRetries)
+}
+
+func IsWireguardKeyProgrammedForTyphaNode(
+	configParams *config.Config,
+	typhaNodeName string,
+	getWireguardHandle func() (netlinkshim.Wireguard, error),
+	calicoClient clientv3.Interface,
+) (bool, error) {
+	logCtx := log.WithField("typhaNodeName", typhaNodeName)
+
+	if !configParams.WireguardHostEncryptionEnabled {
+		// HostEncryption is currently enabled in environments by operator rather than through FelixConfiguration.
+		// This should not change for a given deployment. We only need to handle bootstrapping for clusters where
+		// host encryption is enabled.
+		logCtx.Debug("Host encryption is not enabled - no wireguard bootstrapping required")
+		return true, nil
+	} else if !configParams.WireguardEnabled {
+		// Wireguard is not enabled and so we should have already removed all wireguard config.
+		logCtx.Debug("Wireguard is not enabled - no typha wireguard keys to check")
+		return true, nil
+	}
+
+	logCtx.Debug("Checking typha node wireguard configuration")
+
 	expBackoffMgr := wait.NewExponentialBackoffManager(
-		backoffDuration, backoffMax, time.Minute, backoffExpFactor, jitter, clock.RealClock{})
+		bootstrapBackoffDuration,
+		bootstrapBackoffMax,
+		time.Minute,
+		bootstrapBackoffExpFactor,
+		bootstrapJitter,
+		clock.RealClock{},
+	)
 	defer expBackoffMgr.Backoff().Stop()
 
+	// Make a few attempts to obtain the public key from the datastore for the typha node.
+	var err error
+	var typhaNode *apiv3.Node
+	for r := 0; r < bootstrapMaxRetries; r++ {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		typhaNode, err = calicoClient.Nodes().Get(ctx, typhaNodeName, options.GetOptions{})
```

Comment:
there's a possibilty for this to be blank if a typha override is supplied, perhaps there needs to be handling for this?

Reply:
Oh yuck.  I added the check for `NodeName != nil` in the daemon.go loop which in hindsight is just terrible.   I will fix that up.